### PR TITLE
Support arbitrary mutable queue capacity

### PR DESCRIPTION
**Pull request by artempyanykh**
**TL;DR:** Resolves a confusion around pow 2 capacity of `RingBuffer`. 

Now `RingBuffer.build` and consequently `MutableConcurrentQueue.bounded` return an instance of either `RingBufferPow2` or `RingBufferArb` depending on requested capacity.

To get the best performance a user is expected to use **only** pow 2 capacity queues throughout their codebase.

---------------

**Show me the numbers**

In a regular MPMC benchmark the `RingBufferPow2` implementation gives ~10% better perf then `RingBufferArb` -- JIT optimizes everything nicely.
```
Benchmark                                            (qCapacity)         (qType)   Mode  Cnt           Score   Error   Units
ProducerConsumerBenchmark.Group1SPSC                       65536  RingBufferPow2  thrpt   10           8.968 ± 0.042  ops/us
ProducerConsumerBenchmark.Group1SPSC                       65536   RingBufferArb  thrpt   10           8.370 ± 0.108  ops/us
ProducerConsumerBenchmark.Group1SPSC                       65536         JCTools  thrpt   10           9.011 ± 0.071  ops/us
ProducerConsumerBenchmark.Group1SPSC                       65536     LinkedQueue  thrpt   10           7.624 ± 0.435  ops/us
ProducerConsumerBenchmark.Group1SPSC                       65536     JucBlocking  thrpt   10           5.796 ± 0.809  ops/us

ProducerConsumerBenchmark.Group2ModerateContention         65536  RingBufferPow2  thrpt   10          22.863 ± 1.157  ops/us
ProducerConsumerBenchmark.Group2ModerateContention         65536   RingBufferArb  thrpt   10          20.623 ± 0.293  ops/us
ProducerConsumerBenchmark.Group2ModerateContention         65536         JCTools  thrpt   10          22.848 ± 0.731  ops/us
ProducerConsumerBenchmark.Group2ModerateContention         65536     LinkedQueue  thrpt   10          14.347 ± 0.085  ops/us
ProducerConsumerBenchmark.Group2ModerateContention         65536     JucBlocking  thrpt   10           3.441 ± 2.140  ops/us
```

With a mixed profile (i.e. a mix of unbounded, pow2 and arb `MutableConcurrentQueue` is being used) there is an expected performance hit, but JIT is still able to figure out what's going on and inline things.

(This is a pathological case, in a regular MPMC scenario a performance hit will be much less pronounced)
```
Benchmark                                                    (batchSize)  (qCapacity)  Mode  Cnt   Score   Error  Units
// Only RBPow2 impl is in JIT profile
RingBufferMethodDispatchBenchmark.OnlyPow2                             1            8  avgt   10  24.781 ± 0.673  ns/op

// RBPow2 + LinkedQueue
RingBufferMethodDispatchBenchmark.Pow2Unbounded:puPow2                 1            8  avgt   10  48.626 ± 0.443  ns/op
RingBufferMethodDispatchBenchmark.Pow2Unbounded:puUnbounded            1            8  avgt   10  90.073 ± 1.422  ns/op

// RBPow2 + RBArb + LinkedQueue
RingBufferMethodDispatchBenchmark.All:allPow2                          1            8  avgt   10  54.128 ± 0.431  ns/op
RingBufferMethodDispatchBenchmark.All:allArb                           1            8  avgt   10  49.047 ± 0.612  ns/op
RingBufferMethodDispatchBenchmark.All:allUnbounded                     1            8  avgt   10  97.507 ± 0.618  ns/op
```

**Show me the assembly ᕦ(ò_óˇ)ᕤ**

Below is a greatly shortened version of profiled assembly (MCQ.poll). As you can see the code for `RingBufferPow2` is nicely inlined and on the hot path.

```assembly
........................................
             0x000000011464e7e0: mov    0x8(%rsi),%r10d ; rsi is a pointer to this
             0x000000011464e7e4: shl    $0x3,%r10       ; here we check the class
             0x000000011464e7e8: cmp    %r10,%rax       ; and if it's LinkedQueue
             0x000000011464e7eb: jne    0x0000000114404e20  ; we go to other place
.........................................................................
           [Verified Entry Point]
  1.27%      0x000000011464e800: mov    %eax,-0x14000(%rsp)
  0.03%      0x000000011464e807: push   %rbp
  0.08%      0x000000011464e808: sub    $0x50,%rsp         ;*synchronization entry
.........................................................................
  0.39%      0x000000011464e819: mov    0x90(%rsi),%r11    ;* %r11 = curHead
  0.28%      0x000000011464e820: mov    0x8(%rsi),%ebx     ; read the classword again
  0.08%      0x000000011464e812: mov    0x1a0(%rsi),%r9d   ; getfield capacity
.........................................................................
  0.24%      0x000000011464e830: movslq %r9d,%rdi          ; put capacity into %rdi for RingBufferArb branch
  0.01%      0x000000011464e833: cmp    $0xf8019449,%ebx   ; check the class
         ╭   0x000000011464e839: jne    0x000000011464e90c ; if it's RingBufferArb go to address below
         │   0x000000011464e83f: mov    %r9d,%r10d         ; >>>>>>>>>>>> RingBufferPow2 branch
  0.09%  │   0x000000011464e842: dec    %r10d              ; %r10d = capacity - 1
  0.27%  │   0x000000011464e845: movslq %r10d,%r10
  0.02%  │   0x000000011464e848: and    %r11,%r10          ; curHead & (capacity - 1)
.........................................................................
    the rest of RingBuffer.poll method is here
.........................................................................
  0.06%  │^  0x000000011464e900: add    $0x50,%rsp         ; method epilogue
  0.00%  ││  0x000000011464e904: pop    %rbp
  0.14%  ││  0x000000011464e905: test   %eax,-0xe20a90b(%rip)
         ││                                                ;   {poll_return}
  0.35%  ││  0x000000011464e90b: retq
         ↘│  0x000000011464e90c: cmp    $0xf8019494,%ebx   ; >>>>>>>>>>> RingBufferArb branch
.........................................................................
  0.05%   │  0x000000011464e921: mov    %r11,%rax          ; put curHead into %rax
.........................................................................
  0.04%   │  0x000000011464e93b: cqto                      ; sign extend %rax to %rdx to prepare for idiv
  0.04%   │  0x000000011464e93d: idiv   %rdi
  2.02%   │  0x000000011464e940: mov    %edx,%r10d         ; %edx has the remainder
  0.04%   ╰  0x000000011464e943: jmpq   0x000000011464e84e
```

### CODE REVIEW

**Comment by artempyanykh** [comment]
Tests hung up. Now sure what's going on there.

**Comment by artempyanykh** [comment]
@ghostdogpr sure, I made the change to Queue. As for Travis, hope it'll work this time 😃

**Comment by artempyanykh** [comment]
Thanks John! 
I figured out why test locked up https://github.com/scalaz/scalaz-zio/pull/454/commits/e634694dd3df2b0ec2a70d049fa431942311129c and managed to please CI gods (finally! 😄). Could you take another look?

For the sake of consistency it makes sense to fill the gap and introduce a BoundedQueue of size 1 (a variation on atomic ref or smth).
